### ISSUE TITLE
Failed to open NVENC codec on 1st gen NVENC cards

### ISSUE THREAD

**Issue by vvo1tv** [obsproject/obs-studio#4785]
### Operating System Info

Other

### Other OS

Windows 7

### OBS Studio Version

27.0.0

### OBS Studio Log URL

https://obsproject.com/logs/UT07zHybnD9fbqAZ

### OBS Studio Crash Log URL

https://obsproject.com/logs/OP-ZTgUlPaWKWZ-i

### Expected Behavior

Upgrade to version 27.0.0 and start stream

### Current Behavior

I updated it to version 27.0.0 and could not start the stream with my previous encoder settings. An error appeared - Failed to open NVENC codec. I had to switch from NVEС to x264 Encoder, which increased the load on the CPU.

### Steps to Reproduce

1. Windows 7 x64
2. OBS 27.0.0
3. Settings - Output - Streaming - Encoder - NVENC
...

**Comment by notr1ch** [comment]
The GeForce GTX 670 does not have NVENC, so this is expected and should never have worked.

**Comment by vvo1tv** [comment]
Strange, but in the previous version of the OBS everything worked

**Comment by notr1ch** [comment]
According to https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new the card doesn't support it. If it worked, it was probably experimental at best and might have been removed in a driver update. Windows 7 is also end of life, so you're really in unsupported territory here.

**Comment by guihkx** [comment]
>The GeForce GTX 670 does not have NVENC, so this is expected and should never have worked.

>If it worked, it was probably experimental at best and might have been removed in a driver update. Windows 7 is also end of life, so you're really in unsupported territory here.

Are you sure about that?

I've been using (and still am) NVENC to record with OBS on my GTX 660 since forever. I'm on Linux, though.

The 27.0.0 update from OBS also broke NVENC for me, but I got it working again by switching Output Mode to Advanced and disabling "Psycho Visual Tuning" on the NVIDIA NVENC H.264 encoder.

Here's the full output of OBS 27.0.0 (Flatpak) not being able to use NVENC:

```
$ com.obsproject.Studio
Attempted path: share/obs/obs-studio/locale/en-US.ini
Attempted path: /app/share/obs/obs-studio/locale/en-US.ini
Attempted path: share/obs/obs-studio/locale.ini
Attempted path: /app/share/obs/obs-studio/locale.ini
Attempted path: share/obs/obs-studio/themes/Dark.qss
Attempted path: /app/share/obs/obs-studio/themes/Dark.qss
info: CPU Name: Intel(R) Core(TM) i5-3570K CPU @ 3.40GHz
info: CPU Speed: 3392.389MHz
info: Physical Cores: 4, Logical Cores: 4
info: Physical Memory: 7897MB Total, 4171MB Free
info: Kernel Version: Linux 5.12.8-161-tkg-pds
info: Distribution: "KDE Flatpak runtime" "5.15"
info: Session Type: x11
info: Window System: X11.0, Vendor: The X.Org Foundation, Version: 1.20.11
info: Portable mode: false
Attempted path: share/obs/obs-studio/themes/Dark/no_sources.svg
Attempted path: /app/share/obs/obs-studio/themes/Dark/no_sources.svg
QMetaObject::connectSlotsByName: No matching signal for on_tbar_position_valueChanged(int)
QMetaObject::connectSlotsByName: No matching signal for on_actionShowTransitionProperties_triggered()
QMetaObject::connectSlotsByName: No matching signal for on_actionHideTransitionProperties_triggered()
info: OBS 27.0.0 (linux)
info: ---------------------------------
info: ---------------------------------
info: audio settings reset:
	samples per sec: 48000
	speakers:        2
info: ---------------------------------
info: Initializing OpenGL...
info: Loading up OpenGL on adapter NVIDIA Corporation NVIDIA GeForce GTX 660/PCIe/SSE2
info: OpenGL loaded successfully, version 3.3.0 NVIDIA 465.31, shading language 3.30 NVIDIA via Cg compiler
info: ---------------------------------
info: video settings reset:
	base resolution:   1920x1080
	output resolution: 1920x1080
	downscale filter:  Bicubic
	fps:               60/1
	format:            NV12
	YUV mode:          709/Partial
info: NV12 texture support not available
info: Audio monitoring device:
	name: Default
	id: default
info: ---------------------------------
warning: Failed to load 'en-US' text for module: 'decklink-captions.so'
warning: Failed to load 'en-US' text for module: 'decklink-ouput-ui.so'
libDeckLinkAPI.so: cannot open shared object file: No such file or directory
warning: A DeckLink iterator could not be created.  The DeckLink drivers may not be installed
info: No blackmagic support
warning: v4l2loopback not installed, virtual camera disabled
info: [obs-browser]: Version 2.14.3
info: [obs-browser]: CEF Version 87.1.12+g03f9336+chromium-87.0.4280.88
info: NVENC supported
info: FFMPEG VAAPI supported
error: os_dlopen(/app//lib/obs-plugins/obs-ndi.so->/app//lib/obs-plugins/obs-ndi.so): /app//lib/obs-plugins/obs-ndi.so: cannot open shared object file: No such file or directory

warning: Module '/app//lib/obs-plugins/obs-ndi.so' not loaded
error: os_dlopen(/app//lib/obs-plugins/obs-v4l2sink.so->/app//lib/obs-plugins/obs-v4l2sink.so): /app//lib/obs-plugins/obs-v4l2sink.so: cannot open shared object file: No such file or directory

warning: Module '/app//lib/obs-plugins/obs-v4l2sink.so' not loaded
info: ---------------------------------
info:   Loaded Modules:
info:     text-freetype2.so
info:     rtmp-services.so
info:     obs-x264.so
info:     obs-vst.so
info:     obs-transitions.so
info:     obs-outputs.so
info:     obs-libfdk.so
info:     obs-filters.so
info:     obs-ffmpeg.so
info:     obs-browser.so
info:     linux-v4l2.so
info:     linux-pulseaudio.so
info:     linux-jack.so
info:     linux-decklink.so
info:     linux-capture.so
info:     image-source.so
info:     frontend-tools.so
info:     decklink-ouput-ui.so
info:     decklink-captions.so
info: ---------------------------------
info: ==== Startup complete ===============================================
info: All scene data cleared
info: ------------------------------------------------
info: pulse-input: Server name: 'PulseAudio (on PipeWire 0.3.29) 14.0.0'
info: pulse-input: Audio format: s32le, 48000 Hz, 2 channels
info: pulse-input: Started recording from 'alsa_output.pci-0000_00_1b.0.analog-stereo.monitor'
info: [Loaded global audio device]: 'Desktop Audio'
info: pulse-input: Server name: 'PulseAudio (on PipeWire 0.3.29) 14.0.0'
info: pulse-input: Audio format: s32le, 48000 Hz, 2 channels
info: pulse-input: Started recording from 'alsa_input.pci-0000_00_1b.0.analog-stereo'
info: [Loaded global audio device]: 'Mic/Aux'
info: xshm-input: Geometry 1920x1080 @ 0,0
info: Switched to scene 'Scene'
info: ------------------------------------------------
info: Loaded scenes:
info: - scene 'Scene':
info:     - source: 'Screen Capture (XSHM)' (xshm_input)
info: ------------------------------------------------
Attempted path: share/obs/obs-studio/images/overflow.png
Attempted path: /app/share/obs/obs-studio/images/overflow.png
info: adding 42 milliseconds of audio buffering, total audio buffering is now 42 milliseconds (source: Desktop Audio)

info: ---------------------------------
info: [NVENC encoder: 'simple_h264_recording'] settings:
	rate_control: CQP
	bitrate:      0
	cqp:          16
	keyint:       250
	preset:       hq
	profile:      high
	width:        1920
	height:       1080
	2-pass:       false
	b-frames:     2
	psycho-aq:    1
	GPU:          0

[h264_nvenc @ 0x55924c70c740] Temporal AQ not supported
[h264_nvenc @ 0x55924c70c740] No capable devices found
warning: [NVENC encoder: 'simple_h264_recording'] Failed to open NVENC codec: Generic error in an external library
```

Perhaps there should be a way to detect if the GPU supports such features before automatically enabling it?

**Comment by derrod** [comment]
> According to https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new the card doesn't support it. If it worked, it was probably experimental at best and might have been removed in a driver update. Windows 7 is also end of life, so you're really in unsupported territory here.

The 600 series of GPUs was the first to support NVENC (Kepler), if you check the older NVENC Matrix it does list the 660-690 as supported: https://web.archive.org/web/20200405020102/https://developer.nvidia.com/video-encode-decode-gpu-support-matrix

While support for Kepler (at least consumer) will be dropped with the next major Nvidia driver version it does appear that they were a bit premature with removing Kepler cards from their list here.

**Comment by notr1ch** [comment]
Reopening this as it does seem we have broken compatibility with the oldest GPUs. Even if NVIDIA no longer support them, if they used to work it should still work.

**Comment by RytoEX** [comment]
NVIDIA periodically drops older cards off of its support matrix for NVENC/NVDEC, even if they do have the required hardware.  This makes it appear, at first glance to current information sources, that Kepler GPUs do not support NVENC, so we responded as such.

As for why you @vvo1tv and @guihkx encountered this encoder failure, I'll attempt to clarify what happened on the OBS side of things.  I hope you'll forgive me for this lengthy explanation.  I'll provide a summary at the start for those who don't want to read all the details.


tl;dr:  We recently changed something that exposed Psycho Visual Tuning to all users and it was enabled by default.  It doesn't appear to work on Kepler GPUs, hence the encoder failure.  Disabling Psycho Visual Tuning should allow you to use NVENC.

----

Psycho Visual Tuning (Psycho AQ) has been enabled by default in the [texture-based NVENC implementation](https://github.com/obsproject/obs-studio/commit/ed0c7bcd6a7a7ad844975beda5ec72aa9cc8fcf4) ("jim-nvenc" or "NVIDIA NVENC H.264 (new)") since we introduced the feature on Windows versions newer than Windows 7.  We initially [enabled the setting by default in the FFmpeg NVENC implementation](https://github.com/obsproject/obs-studio/commit/8b566f3352bf7a60fc490451f2a4c1d36b328dff#diff-fc007c567208a25eac997f8ad710956b8be02bbc3f95b401719c971830ca42ebR406).  However, this caused some issues, so we [disabled it temporarily](https://github.com/obsproject/obs-studio/commit/ecdae0d4175309e1740dc030c69faf7aa8836baa).  After we [tied the setting to the texture-based implementation](https://github.com/obsproject/obs-studio/commit/e49bba3292a35cc8ce87911ad54627de1e242b20#diff-fc007c567208a25eac997f8ad710956b8be02bbc3f95b401719c971830ca42ebR490-R497), we then [re-enabled it "temporarily"](https://github.com/obsproject/obs-studio/commit/e67d04270aad0e00ed154805ac6619c42ede25ee), and we no longer passed it to FFmpeg.

To streamline the user experience, we allowed OBS to fall back from the texture-based implementation to the FFmpeg NVENC implementation if necessary (e.g., the user enabled scaling in the encoder settings).  In a recent effort to minimize differences in our two NVENC implementations, we merged #4344, which allowed OBS to read the setting for Psycho Visual Tuning and pass it to the FFmpeg NVENC implementation.  As I mentioned earlier, this setting was now enabled by default in both implementations.  However, this exposed the issue you ran into, which is that Kepler GPUs do not support Psycho Visual Tuning (I'm not sure that Kepler GPUs support any Adaptive Quantization).  On systems that only have the FFmpeg NVENC implementation available (Windows 7 and Linux users), this option was not shown in the UI, so users could not disable it.  In response, we merged #4477 to expose the option in the UI so that affected users could disable it, but we continued to leave it enabled by default as most users should benefit from it.


> Perhaps there should be a way to detect if the GPU supports such features before automatically enabling it?

Incidentally, in the textured-based implementation, we _do_ check if the GPU supports this feature before trying to enable it.

https://github.com/obsproject/obs-studio/blob/27.0.0/plugins/obs-ffmpeg/jim-nvenc.c#L506-L511

I recently took a look at adding this check to our FFmpeg NVENC implementation, but it required shuffling a fair bit of code around to be able to automatically detect this, and I didn't get it working.  Someone else may have to tackle that.  However, with the number of Kepler GPUs actively used in the wild compared to newer GPUs, and with NVIDIA expected to shift Kepler GPUs from having active driver support to LTS possibly by the end of this year, I can't say that it'll be high priority since users can disable the options causing the encoder failure.

**Comment by guihkx** [comment]
Thanks a lot for the insight. And no worries, I completely understand ancient GPUs like mine having lower priority. ^^

**Comment by RytoEX** [comment]
For those wishing to test a fix, please feel free to test a build from #4825.
